### DISCUSSION TITLE
Packaging script

### DISCUSSION THREAD

**Discussion by uulm-janbaudisch** [msys2/MSYS2-packages#4290] [Ideas]
I was working on getting some software to build on Windows and ended up using MSYS2. While the build process worked, I was not sure on how to make it distributable. I came across [this script](https://sourceforge.net/p/msys2/tickets/34/#1cb7) which is also linked in the documentation. I guess there is no "official" tool to create an archive so that the binaries in it can be used in a non-MSYS2 environment?
For said project I built on the mentioned script and made it a little bit more generic: https://gist.github.com/uulm-janbaudisch/6746d55083c1b13b7dce48bbfb771863
So I was wondering if this script could be of some use for the MSYS2 project (for example by including it) or whether there are alternatives?

**Comment by MehdiChinoune** [comment]
Sometimes copying dynamic libraries is sufficient, but It is not always.
Some libraries depends on some plugins, data files ...etc to run. They should be copied as well and they must be placed in the same relative path (share, lib, etc ...).
Some apps do not link directly to libraries but load them at runtime, there is no tool to detect that.
The safest method is to copy the packages with all their files, but that results to a huge size of the archive.

**Reply by sskras** [reply]
> Some apps do not link directly to libraries but load them at runtime, there is no tool to detect that.

Actually I think `strace` in combination with simulated run could be used to detect all successfully opened files.

**Reply by uulm-janbaudisch** [reply]
I am not very familiar with shipping software on Windows but I guess the copy-dynamic-libraries approach could cover most use cases? Maybe, one could add some extra options to include or exclude directories from packages, for example to copy `bin`, `lib` and `etc`. Sure, it won't cover all use cases, but so far I couldn't find any solution to begin with.

Another idea would be to also provide the option to declare runtime dependencies to always include, even if not detected as a dependency.
